### PR TITLE
Server client connection

### PR DESCRIPTION
**Pull request by fi-ya**
This pull request refers to this [trello](https://trello.com/b/g7UHpv0V/apprenticeship) and includes the following:
- server socket created successfully
- client socket connection successfully
- client socket input/output streams handled using BufferedReader/PrintWriter
- used netcat to get client request
- client response sent successfully

To run [acceptance test](https://github.com/8thlight/http_server_spec/tree/master/features/01_getting_started), run program `./gradlew run`, cd into http_server_spec and run `rake test:f1 # Run all of the tests in 01_getting_started` 
**Acceptance test :** 
- [x] pass for Simple GET & Simple HEAD 
(21) Successful, (0) Pending, (0) Undefined, (4) Failed, (2) Error

- [x] circleci build & test passing 

Thank you for your feedback!

### CODE REVIEW

File: src/main/java/org/httpserver/request/RequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    public String responseBuilder(){
+        statusCode = "200";
+        statusText = "OK";
+        responseStatusLine = httpVersion + space + statusCode + space + statusText + CRLF;
+
+        if(Objects.equals(requestTarget, "/simple_get")){
+            responseBody = "";
+        } else {
+            if (Objects.equals(requestTarget, "/simple_get_with_body")) {
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/request/RequestHandler.java:38]: You could make this an else if, removing one level of indentation

File: src/main/java/org/httpserver/request/RequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                responseBody = "Hello world";
+            }
+        }
+        response = responseStatusLine + CRLF + responseBody;
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/request/RequestHandler.java:42]: Nice use of a constant here (CRLF). You could do something similar with the Status code and test perhaps..

File: src/main/java/org/httpserver/request/RequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package org.httpserver.request;
+
+import java.util.Objects;
+
+public class RequestHandler {
+    private final String clientRequest;
+    String httpMethod;
+    String requestTarget;
+    String httpVersion ;
+    String CRLF = "\r\n";
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/request/RequestHandler.java:10]: Try to make this a constant. This is just a variable currently. The convention in java is for variables to be lowercase and constants to be UPPERCASE.

File: src/main/java/org/httpserver/request/RequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    String space = " ";
+    String statusCode;
+    String statusText;
+    String responseStatusLine;
+    String responseBody;
+    String response;
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/request/RequestHandler.java:16]: You may not need these variables to be "global" across this class. If you only use one of these variables in a method below and nowhere else, it maybe better to declare it in there. Keep declarations and usage of variables close to each other if you can.

Make variables a property of a class if it's necessary.

File: src/main/java/org/httpserver/request/RequestHandler.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+
+    public RequestHandler(String clientRequest){
+        this.clientRequest = clientRequest;
+    }
+
+    public void parseClientRequest(){
+        String[] arrOfSplitResponseStr = clientRequest.split(" ");
+        this.httpMethod = arrOfSplitResponseStr[0];
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/request/RequestHandler.java:25]: What happens here if the client request returns an empty array?

File: src/main/java/org/httpserver/request/RequestHandler.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        this.requestTarget = arrOfSplitResponseStr[1];
+        this.httpVersion = arrOfSplitResponseStr[2];
+    }
+
+    public String responseBuilder(){
+        statusCode = "200";
+        statusText = "OK";
+        responseStatusLine = httpVersion + space + statusCode + space + statusText + CRLF;
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/request/RequestHandler.java:33]: As per my comment above, you can probably declare responseStatusLine here as I think it is only used in this method.

File: src/main/java/org/httpserver/server/Server.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import java.net.ServerSocket;
+import java.net.Socket;
+
+public class Server {
+    private final int portNumber;
+
+    public Server(int portNumber){
+        this.portNumber = portNumber;
+    }
+    public void start() throws IOException {
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/server/Server.java:16]: Put an empty line above start. I know it's minor but I'm a stickler for blank lines :-)

File: src/main/java/org/httpserver/server/Server.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+            String clientRequest = clientHandler.processRequest();
+
+            RequestHandler requestHandler = new RequestHandler(clientRequest);
+            requestHandler.parseClientRequest();
+
+            String response = requestHandler.responseBuilder();
+
+            clientHandler.processSendResponse(response);
+            clientHandler.closeClientConnection();
+        }
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/server/Server.java:35]: I think you may be able to abstract some of the responsibilities in the loop as we discussed. It's looking on the right tracks though

File: src/main/java/org/httpserver/server/ServerWrapper.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    public ServerSocket createServerSocket(int portNumber) throws IOException {
+        ServerSocket serverSocket = null;
+        try{
+            serverSocket = new ServerSocket(portNumber);
+            serverLogger.listeningForClientRequest(portNumber);
+        } catch (IOException ioException){
+            ioException.printStackTrace();
+            serverLogger.printFailedToCreateServerSocket(portNumber);
+            System.exit(1);
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/server/ServerWrapper.java:22]: You may want to think about replacing printStackTrace and System.out.println with logback or something similar. In this particular section, this is a serious fatal error. You could replace lines 20 and 21 with something like Logback's log.fatal. That maybe the incorrect usage though. Generally we should avoid System.out and instead logging frameworks. Just not log4j!!! :-D

File: src/main/java/org/httpserver/server/ServerWrapper.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+    public Socket createClientSocket(ServerSocket serverSocket)throws IOException {
+        Socket clientSocket = null;
+        try{
+            clientSocket = serverSocket.accept();
+            serverLogger.printConnectedClientSocket(clientSocket.getPort());
+        } catch(IOException ioException){
+            ioException.printStackTrace();
+            serverLogger.printFailedClientSocketConnection();
+        }
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/server/ServerWrapper.java:35]: I would replace these 2 print statements with a call to a logger, as previously mentioned.

File: src/main/java/org/httpserver/server/StdOutServerLogger.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package org.httpserver.server;
+
+import java.util.Date;
+
+public class StdOutServerLogger implements ServerLogger {
+
+    public void listeningForClientRequest(int portNumber){
+        System.out.println("[+] Server socket connection created. (" + new Date() + "). Listening on port " + portNumber);
+    };
+    public void printFailedToCreateServerSocket(int portNumber){
```
Comment (daddyshortlegs) [inline src/main/java/org/httpserver/server/StdOutServerLogger.java:10]: Blank lines between methods please :-)

**Review by daddyshortlegs — CHANGES_REQUESTED** [review]
Good set of changes and nice to see you've removed the .idea folder. 
A nice feature of IntelliJ is the ability to auto format your code. I believe with Mac you can do command/option/L to do it, that would sort out the spaces and blank lines that I'm a stickler for.
I mentioned "global" variables (properties of the class). Try to avoid this if you can as this is effectively "state" and it's generally a good idea to avoid state whenever possible, especially when dealing with multi-threaded programs, like this one would become. You also don't need to declare your variables so far away from where they are used. The "scope" of your variables should be small.
Looks like your making good progress though.

**Review by devlinmorrow — APPROVED** [review]
I think this is a good PR which shows that you have developed your understanding of the setup of a server.

File: src/main/java/org/httpserver/request/RequestHandler.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+        if(Objects.equals(requestTarget, "/simple_get")){
```
[the remark this answers is not in the document]
Comment (fi-ya) [reply src/main/java/org/httpserver/request/RequestHandler.java:35]: IntelliJ suggested refactoring as 'String values are compared using '==', not 'equals()'
Screenshot 2022-05-30 at 10 01 55

File: src/main/java/org/httpserver/request/RequestHandler.java (hunk 1 above)
Comment (fi-ya) [reply src/main/java/org/httpserver/request/RequestHandler.java:38]: Yes, that works.

File: src/main/java/org/httpserver/request/RequestHandler.java (hunk 4 above)
Comment (fi-ya) [reply src/main/java/org/httpserver/request/RequestHandler.java:16]: Thanks, I wasn't sure if i should make them attributes of the class, but i was thinking of the next possible uses and so made them global. But i have refactored and make them scoped to the method they are being used in.

File: src/main/java/org/httpserver/request/RequestHandler.java (hunk 5 above)
Comment (fi-ya) [reply src/main/java/org/httpserver/request/RequestHandler.java:25]: It errors out. I will add an if statement which handles this case. (breaks out of if statement)

File: src/main/java/org/httpserver/request/RequestHandler.java (hunk 6 above)
Comment (fi-ya) [reply src/main/java/org/httpserver/request/RequestHandler.java:33]: I have removed all unessarary global variables.

File: src/main/java/org/httpserver/server/ServerWrapper.java (hunk 9 above)
Comment (fi-ya) [reply src/main/java/org/httpserver/server/ServerWrapper.java:22]: Great, thanks. I'll look into how to implement Logback in a Gradle project.

File: src/main/java/org/httpserver/server/StdOutServerLogger.java (hunk 11 above)
Comment (fi-ya) [reply src/main/java/org/httpserver/server/StdOutServerLogger.java:10]: Command +Shift +L is fast becoming my fav shortcut! :)

File: src/test/java/org/httpserver/server/ServerWrapperTest.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        ServerSocket newServerSocket =  serverWrapper.createServerSocket(5000);
+        assertNotNull(newServerSocket);
+    }
+
+    @Test
+    void serverSocketFailsToConnectToInvalidPort() throws IOException {
+        StdOutServerLogger serverLogger = new StdOutServerLogger();
+        ServerWrapper serverWrapper = new ServerWrapper(serverLogger);
+
+        assertThrows(IllegalArgumentException.class, () -> serverWrapper.createServerSocket(-1));
```
[the remark this answers is not in the document]
Comment (fi-ya) [reply src/test/java/org/httpserver/server/ServerWrapperTest.java:27]: Thanks!

File: src/main/java/org/httpserver/client/ClientHandler.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    public String processRequest() throws IOException {
+        BufferedReader clientRequestReader = createClientSocketInputStream();
+        return getClientRequest(clientRequestReader);
+    }
+
+    public BufferedReader createClientSocketInputStream() {
+        try {
+            clientConnectionCounter++;
```
[the remark this answers is not in the document]
Comment (fi-ya) [reply src/main/java/org/httpserver/client/ClientHandler.java:32]: Thanks I will include this test in the next PR

**Pull request merged**